### PR TITLE
[ENH] New version of Random Dot Motion stimulu

### PR DESCRIPTION
**Pull request by Gilles86**
I made some new Random Dot Motion stimuli that are often used in the literature. I'm curious what you think.

### CODE REVIEW

File: psychopy/visual/dot.py [hunk 1]
```diff
@@ -504,3 +504,199 @@ def _update_dotsXY(self):
 
         # update the pixel XY coordinates in pixels (using _BaseVisual class)
         self._updateVertices()
+
+
+
+
```
Comment (hoechenberger) [inline psychopy/visual/dot.py]: Please only insert 2 blank lines here.

File: psychopy/visual/dot.py [hunk 2]
```diff
@@ -504,3 +504,199 @@ def _update_dotsXY(self):
...
+    def coherence(self, coherence):   
+        self._signalDots = numpy.ones(self.nDots, dtype=bool)
+
+    @attributeSetter
+    def dir(self, dir):
+        self.__dict__['dir'] = dir / 180 * numpy.pi
+        self._dotsDir = numpy.random.vonmises(self.dir, self.kappa, self.nDots)
+    
+    @attributeSetter
+    def kappa(self, kappa):
```
Comment (hoechenberger) [inline psychopy/visual/dot.py]: Please document the parameter.

File: psychopy/visual/dot.py [hunk 3]
```diff
@@ -504,3 +504,199 @@ def _update_dotsXY(self):
...
+    def __init__(self, screen, kappa=1, *args, **kwargs):
+        self.__dict__['kappa'] = kappa
+        super(VonMisesDotStim, self).__init__(screen, *args, **kwargs)
+        
+    @attributeSetter
+    def coherence(self, coherence):   
+        self._signalDots = numpy.ones(self.nDots, dtype=bool)
+
+    @attributeSetter
+    def dir(self, dir):
```
Comment (hoechenberger) [inline psychopy/visual/dot.py]: Please document the parameter.

File: psychopy/visual/dot.py [hunk 4]
```diff
@@ -504,3 +504,199 @@ def _update_dotsXY(self):
...
+        self.__dict__['dir'] = dir / 180 * numpy.pi
+        self._dotsDir = numpy.random.vonmises(self.dir, self.kappa, self.nDots)
+    
+    @attributeSetter
+    def kappa(self, kappa):
+        self.kappa = kappa
+        self._dotsDir = numpy.random.vonmises(self.dir, self.kappa, self.nDots)
+
+    def _update_dotsXY(self):
+
```
Comment (hoechenberger) [inline psychopy/visual/dot.py]: Superfluous blank line.

File: psychopy/visual/dot.py [hunk 5]
```diff
@@ -504,3 +504,199 @@ def _update_dotsXY(self):
...
+        # Get dead dots
+        if self.dotLife > 0:  # if less than zero ignore it
+            # decrement. Then dots to be reborn will be negative
+            self._dotsLife -= 1
+            dead = (self._dotsLife <= 0.0)
+            self._dotsLife[dead] = self.dotLife
+        else:
+            dead = numpy.zeros(self.nDots, dtype=bool)
+
+
```
Comment (hoechenberger) [inline psychopy/visual/dot.py]: Superfluous blank line.

File: psychopy/visual/dot.py [hunk 6]
```diff
@@ -504,3 +504,199 @@ def _update_dotsXY(self):
...
+        else:
+            dead = numpy.zeros(self.nDots, dtype=bool)
+
+        
+        # update dead dots
+        if sum(dead):
+            self._verticesBase[dead, :] = self._newDotsXY(sum(dead))
+            self._dotsDir[dead] = numpy.random.vonmises(self.dir, self.kappa, sum(dead))
+
+
```
Comment (hoechenberger) [inline psychopy/visual/dot.py]: Superfluous blank line.

File: psychopy/visual/dot.py [hunk 7]
```diff
@@ -504,3 +504,199 @@ def _update_dotsXY(self):
...
+        self._verticesBase[:, 0] += self.speed * cosDots
+        self._verticesBase[:, 1] += self.speed * sinDots
+
+        # find Out-of-bounds dots
+        if self.fieldShape in (None, 'square', 'sqr'):
+            out0 = (numpy.abs(self._verticesBase[:, 0]) > 0.5*self.fieldSize[0])
+            out1 = (numpy.abs(self._verticesBase[:, 1]) > 0.5*self.fieldSize[1])
+            outofbounds = out0 + out1
+
```
Comment (hoechenberger) [inline psychopy/visual/dot.py]: Superfluous blank line.

File: psychopy/visual/dot.py [hunk 8]
```diff
@@ -504,3 +504,199 @@ def _update_dotsXY(self):
...
+        elif self.fieldShape == 'circle':
+            normXY = self._verticesBase / 0.5 / self.fieldSize
+            outofbounds = (numpy.hypot(normXY[:, 0], normXY[:, 1]) > 1)   
+
+        # update any out of bound dots dots
+        if sum(outofbounds):
+            self._verticesBase[outofbounds, :] = self._newDotsXY(sum(outofbounds))
+            self._dotsDir[outofbounds] = numpy.random.vonmises(self.dir, self.kappa, sum(outofbounds))
```
Comment (hoechenberger) [inline psychopy/visual/dot.py]: Line is too long, please try to limit yourself to 79 chars if possible.

File: psychopy/visual/dot.py [hunk 9]
```diff
@@ -504,3 +504,199 @@ def _update_dotsXY(self):
...
+
+        # update the pixel XY coordinates in pixels (using _BaseVisual class)
+        self._updateVertices()
+
+
+class GoldShadlenDotStim(DotStim):
+
+    ''' This version of the Random Dot Motion-task is the one first described
```
Comment (hoechenberger) [inline psychopy/visual/dot.py]: Please use triple-double-quotes.

File: psychopy/visual/dot.py [hunk 10]
```diff
@@ -504,3 +504,199 @@ def _update_dotsXY(self):
...
+     the representation of a perceptual decision in developing oculomotor commands.
+     The Journal of Neuroscience, 23(2), 632–651.
+     * Forstmann, B. U., Dutilh, G., Brown, S., Neumann, J., Cramon, Von, D. Y.,
+     Ridderinkhof, K. R., & Wagenmakers, E.-J. (2008). Striatum and pre-SMA facilitate
+     decision-making under time pressure. Proc Natl Acad Sci USA, 105(45), 17538–17542.
+     * van Maanen, L., Brown, S. D., Eichele, T., Wagenmakers, E.-J., Ho, T.,
+     Serences, J., & Forstmann, B. U. (2011). Neural correlates of trial-to-trial
+     fluctuations in response caution. The Journal of Neuroscience, 31(48), 17488–17495.
+     http://doi.org/10.1523/JNEUROSCI.2924-11.2011
+     '''
```
Comment (hoechenberger) [inline psychopy/visual/dot.py]: Please use triple-double-quotes.

File: psychopy/visual/dot.py [hunk 11]
```diff
@@ -504,3 +504,199 @@ def _update_dotsXY(self):
...
+
+    def __init__(self, 
+                 win,
+                 nFrames=3,
+                 signalDots='different',
+                 noiseDots='walk',
+                 fieldShape='circle',
+                 dotLife=0,
+                 nDots=500,
+                 *args, **kwargs):
```
Comment (hoechenberger) [inline psychopy/visual/dot.py]: Please document all parameters, and explicitly list parameters that are going to be passed to the parent class.

File: psychopy/visual/dot.py [hunk 12]
```diff
@@ -504,3 +504,199 @@ def _update_dotsXY(self):
...
+    def coherence(self, coherence):
+        for dotstim in self.dotstims:
+            dotstim.coherence = coherence
+
+    @attributeSetter
+    def dir(self, dir):
+        for dotstim in self.dotstims:
+            dotstim.dir = dir
+
+
```
Comment (hoechenberger) [inline psychopy/visual/dot.py]: Superfluous blank line.

**Comment by Gilles86** [comment]
Made some quick fixes.

Shouldn't we refactor this code a bit and/or use the elementArrayStim to make it faster?

**Comment by hoechenberger** [comment]
Umm I'm not sure whether it was such a great idea to merge my yet-unmerged branch into yours! The better approach probably would have been to wait for it to get merged into master, and then rebase your branch onto master. Or to rebase your branch onto mine, if you don't want to wait for it to get merged. I'm a bit concerned this might mess up things now, but let's see…

**Comment by hoechenberger** [comment]
> Shouldn't we refactor this code a bit and/or use the elementArrayStim to make it faster?

I believe @peircej would be the expert for this question :)

**Comment by Gilles86** [comment]
Sorry about the merge. I'm trying to learn git and got a bit too enthusiastic. 

I'll leave this request alone for now.

**Comment by hoechenberger** [comment]
> Sorry about the merge. I'm trying to learn git and got a bit too enthusiastic.

Hehe don't worry, `git` is quite difficult to handle at the beginning. Don't ask how many times I entirely destroyed my repositories and lost commits! I still do regular backups before "invasive" git operations…

**Comment by peircej** [comment]
In terms of refactoring, I've been wondering about whether we should be separating the notion of RDK motion sequences from the stimulus altogether. i.e. the code that generates the XY positions should not be the stimulus itself, but a distinct class that can, indeed act as the input for something like ElementArray.
As to whether we should add this additional RDK method I don't know about that either! It is true that the noise Shadlen uses is the most common, but it just doesn't seem as *general* as the Scase et al formulation. But Scase et al options do see to confuse people!

**Comment by Gilles86** [comment]
@peircej, thanks for your input!

I talked to multiple people that "just want to replicate the RDM stimulus" of earlier papers. The Shadlen-RDK is very popular in perceptual decision-making research using sequential sampling models (DDM/LBA).

As I see it, the Shadlen-stimulus really is just an extension of the Scase et al. formulation, where you loop over n (usually 3) frames. 

I personally like the von Mises-RDK more because it's so much more easy on the eyes.

I guess it's up to you as maintainers of psychopy whether you want a package that can provide only elementary stimuli to build up experiments, or also more ready-made more complex stimuli. I know people are interested in the latter, but maybe those should be part of a meta-package.

**Comment by peircej** [comment]
Yes, I'm coming around to the idea of the vonMises and Shadlen pre-canned methods. I guess I'll need to think about how to make this nice for Builder as well. I suppose well do something like in the Loop dialog, where people select the type firs and then set the relevant parameters. So we could have Dot Stim with types [von Mises, Shadlen et al, Scase et al] and if the user selects Scase they get the current set of options, if they choose Shadlen they get fewer options, if the choose von Mises they get different options.

I'd still like some way that we could conveniently use the xy positions from these methods in other stimuli (like element array). The current rendering of dots is limited to whatever the opengl implementation decides (usually that means "dots" are squares), whereas ElementArray would allow circles, gaussians, DoGs, anything. Maybe the way to handle that is a funky hack: 

* users could create a DotStim (or derivative implementation) as they do now and it would render with simple (opengl-implemented) dots
* if the user provides that DotStim as the XYs argument to ElementArrayStim then `DotStim.draw()` is not used, but ElementArrayStim updates the dots XYs during _it's_ `draw()`

It's neat, but maybe too obscure?!

**Comment by Gilles86** [comment]
Sounds good.

Or what about a DotStim-object that takes an ElementArrayStim and DotStim takes care of the positions from frame to frame, depending on its 'method'. And by default it takes an ElementArrayStim with little dots? That seems more natural to me. In such a model, the DotStim-object is the one taking care of repositioning and making new dots, whereas the ElementArrayStim takes care of making a group of stimuli and their appearance. 

But maybe you are worried about backwards compatibility?

One other thing to think about is something Martin (Szinte) pointed me to: a problem with the original RDKs consisting of white dots on a black background is that your visual system adapts and the dots sometimes seem to disappear. In some of his papers they tried to increase contrast by always using both black _and_ white dots on a grey background. It would be cool if we could also make this variant easy to make with psychopy. Maybe ElementArrayStim could do this for us?

**Comment by peircej** [comment]
I left this open for ages thinkin we should still look at it one day. I still want to revisit the whole issue one day, but the day hasn't come yet!

**Pull request closed without merging**